src/devices_models/devices/hydro_generation.jl

```diff
@@ -434,97 +435,18 @@ function initial_conditions!(
-     peak_value_function = x -> PSY.get_rating(x) * sin(acos(PSY.get_powerfactor(x)))
-    else
-     forecast_label = "get_rating"
-     peak_value_function = x -> PSY.get_reactivepower(x)
-    end
-    constraint_data = Vector{DeviceTimeSeries}(undef, length(devices))
-    for (ix, d) in enumerate(devices)
-     ts_vector = get_time_series(psi_container, d, forecast_label)
-     timeseries_data = DeviceTimeSeries(d, peak_value_function, ts_vector)
-     constraint_data[ix] = timeseries_data
-    end
-
-    if parameters
-     include_parameters(
-         psi_container,
-         constraint_data,
-         UpdateRef{R}(REACTIVE_POWER, forecast_label),
-         :nodal_balance_active,
-     )
-     return
-    else
-     for t in model_time_steps(psi_container)
-         for device in constraint_data
-             add_to_expression!(
-                 psi_container.expressions[:nodal_balance_reactive],
-                 device.bus_number,
-                 t,
-                 device.multiplier * device.timeseries[t],
-             )
-         end
-     end
-    end
-    return
-    =#
-end
-
-function nodal_expression!(
-    psi_container::PSIContainer,
-    devices::IS.FlattenIteratorWrapper{H},
-    system_formulation::Type{<:PM.AbstractActivePowerModel},
-) where {H <: PSY.HydroGen}
-    parameters = model_has_parameters(psi_container)
-    use_forecast_data = model_uses_forecasts(psi_container)
-    if use_forecast_data
-        forecast_label = "get_rating"
-        peak_value_function = x -> PSY.get_rating(x)
-    else
-        forecast_label = ""
-        peak_value_function = x -> PSY.get_activepower(x)
-    end
-    constraint_data = Vector{DeviceTimeSeries}(undef, length(devices))
-    for (ix, d) in enumerate(devices)
-        ts_vector = get_time_series(psi_container, d, forecast_label)
-        timeseries_data = DeviceTimeSeries(d, peak_value_function, ts_vector)
-        constraint_data[ix] = timeseries_data
-    end
+########################## Addition to the nodal balances #################################
 
-    if parameters
-        include_parameters(
-            psi_container,
-            constraint_data,
-            UpdateRef{H}(ACTIVE_POWER, forecast_label),
-            :nodal_balance_active,
-        )
-        return
-    else
-        for t in model_time_steps(psi_container)
-            for device in constraint_data
-                add_to_expression!(
-                    psi_container.expressions[:nodal_balance_active],
-                    device.bus_number,
-                    t,
-                    device.multiplier * device.timeseries[t],
-                )
-            end
-        end
-    end
-    return
+function NodalExpressionInputs(
+    ::Type{<:PSY.HydroGen},
+    ::Type{<:PM.AbstractActivePowerModel},
+    use_forecasts::Bool,
+)
+    return NodalExpressionInputs(
+        "get_rating",
+        ACTIVE_POWER,
+        use_forecasts ? x -> PSY.get_rating(x) : x -> PSY.get_activepower(x),
+    )
```

Comment:
This change needs scrutiny.  Note that there is no implementation for reactive power.

Reply:
Point taken, lets addressed it in a separate PR